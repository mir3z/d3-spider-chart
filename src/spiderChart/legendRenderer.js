import * as d3 from "d3";
import noop from "lodash.noop";

/**
 * Renders vertically centered legend at the right side of *anchor*.
 * Clicking each legend cell updates chart data with properties that can be used to toggle series visibility.
 * @param parentGroup SVG parent group
 * @param anchor SVG element being used as a reference point
 * @param color color scale to colorize each legend cell
 * @returns {function}
 */
export default function legendRenderer(parentGroup, { anchor = parentGroup, color }) {
    const group = parentGroup
        .append("g")
        .attr("class", CLASS_NAMES.LEGEND_GROUP);

    const drawSymbol = (data, update) => (d, i, nodes) => {
        const el = d3.select(nodes[i]);

        el.select(`.${ CLASS_NAMES.CELL_SYMBOL }`).remove();

        const symbol = SYMBOL_MAP[d.type](el, color(i), !d.hidden);
        return symbol.on("click", handleClick(data, update));
    };

    const setLabel = (data, update) => el => el
        .text(d => d.key)
        .attr("fill", d => d.hidden ? "#afafaf" : "#000")
        .on("click", handleClick(data, update));

    const setHorizontalBorder = (y) => el => el
        .append("line")
        .attr("class", CLASS_NAMES.LINE_DECOR)
        .attr("x1", 0)
        .attr("y1", y)
        .attr("x2", 300)
        .attr("y2", y);

    const positionLegend = () => el => {
        const refPos = anchor.node().getBoundingClientRect();
        const groupPos = group.node().getBoundingClientRect();

        const x = refPos.right + 50;
        const y = refPos.top + (refPos.height - groupPos.height) / 2;

        return el.attr("transform", `translate(${ x }, ${ y + 12 })`);
    };

    const handleClick = (data, update) => d => {
        const nextData = data
            .map(series => series.key === d.key
                ? { ...series, hidden: !series.hidden }
                : series
            );

        update(nextData);
    };

    return function render(data = [], update = noop) {
        const legendItem = group
            .selectAll(`.${ CLASS_NAMES.LEGEND_CELL }`)
            .data(data);

        legendItem.exit().remove();

        const enterItem = legendItem.enter()
            .append("g")
            .attr("class", CLASS_NAMES.LEGEND_CELL)
            .attr("transform", (d, i) => `translate(0, ${ i * CELL_HEIGHT })`);

        enterItem
            .append("text")
            .attr("class", CLASS_NAMES.CELL_LABEL)
            .attr("alignment-baseline", "central")
            .attr("dominant-baseline", "central")
            .attr("x", "2.5em")
            .call(setLabel(data, update));

        enterItem.each(drawSymbol(data, update));

        legendItem
            .select(`.${ CLASS_NAMES.CELL_LABEL }`)
            .call(setLabel(data, update));

        legendItem.each(drawSymbol(data, update));

        group.selectAll(`.${ CLASS_NAMES.LINE_DECOR }`).remove();
        group
            .call(setHorizontalBorder(-BORDER_SPACING))
            .call(setHorizontalBorder((data.length - 1) * CELL_HEIGHT + BORDER_SPACING))
            .call(positionLegend());

        return group;
    };
}

const CELL_HEIGHT = 30;
const BORDER_SPACING = 20;
const CLASS_NAMES = {
    LEGEND_GROUP: "legend",
    LEGEND_CELL: "legend-cell",
    CELL_LABEL: "legend-cell__label",
    CELL_SYMBOL: "legend-cell__symbol",
    LINE_DECOR: "line-decor"
};

const drawAreaSymbol = (el, color, isActive) => el
    .append("rect")
    .attr("class", CLASS_NAMES.CELL_SYMBOL)
    .attr("fill", isActive ? color : "#fff")
    .attr("stroke", isActive ? "none": "#afafaf")
    .attr("y", "-0.5em")
    .attr("width", "2em")
    .attr("height", "1em");

const drawLineSymbol = (el, color, isActive) => el
    .append("line")
    .attr("class", CLASS_NAMES.CELL_SYMBOL)
    .attr("stroke", isActive ? color : "#afafaf")
    .attr("stroke-width", 4)
    .attr("x1", 0)
    .attr("x2", "2em");

const drawDotSymbol = (el, color, isActive) => el
    .append("circle")
    .attr("class", CLASS_NAMES.CELL_SYMBOL)
    .attr("fill", isActive ? color : "#fff")
    .attr("stroke", isActive ? "none": "#afafaf")
    .attr("r", "4px")
    .attr("cx", "1em");

const SYMBOL_MAP = {
    "area": drawAreaSymbol,
    "line": drawLineSymbol,
    "dot": drawDotSymbol
};
