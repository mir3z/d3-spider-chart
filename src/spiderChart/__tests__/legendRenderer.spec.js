import * as d3 from "d3";
import legendRenderer from "../legendRenderer";
import { cleanup, findElementByText, setup } from "./utils";

describe("legend renderer", () => {
    const color = d3.scaleLinear().domain([0, 1, 2]).range(["#f00", "#0f0", "#00f"]);

    afterEach(cleanup);

    it("renders its own group", () => {
        const svg = setup();
        const renderLegend = legendRenderer(svg, {});

        const group = renderLegend();
        const ownGroup = svg.node().querySelector("g.legend");

        expect(ownGroup).not.toBeNull();
        expect(group.node()).toEqual(ownGroup);
    });

    it("renders correct legend cell for each series", () => {
        const svg = setup();
        const renderLegend = legendRenderer(svg, { color });
        const parent = renderLegend([
            { type: "area", key: "series 1" },
            { type: "line", key: "series 2" },
            { type: "dot", key: "series 3" }
        ]);

        expect(parent.node()).toMatchSnapshot();
    });

    it("renders correct legend cell for each hidden series", () => {
        const svg = setup();
        const renderLegend = legendRenderer(svg, { color });
        const parent = renderLegend([
            { type: "area", key: "series 1", hidden: true },
            { type: "line", key: "series 2", hidden: true },
            { type: "dot", key: "series 3", hidden: true }
        ]);

        expect(parent.node()).toMatchSnapshot();
    });

    describe("when clicking series label", () => {
        it("updates series state with hidden = true if series is visible", () => {
            const svg = setup();
            const renderLegend = legendRenderer(svg, { color });
            const data = [
                { type: "area", key: "series 1" },
                { type: "line", key: "series 2" }
            ];
            const update = jest.fn();
            const parent = renderLegend(data, update);

            findElementByText("series 1", parent.node())
                .dispatchEvent(new Event("click"));

            expect(update).toHaveBeenCalledWith([
                { type: "area", key: "series 1", hidden: true },
                { type: "line", key: "series 2" }
            ]);
        });

        it("updates series state with hidden = false if series is NOT visible", () => {
            const svg = setup();
            const renderLegend = legendRenderer(svg, { color });
            const data = [
                { type: "area", key: "series 1" },
                { type: "line", key: "series 2", hidden: true }
            ];
            const update = jest.fn();
            const parent = renderLegend(data, update);

            findElementByText("series 2", parent.node())
                .dispatchEvent(new Event("click"));

            expect(update).toHaveBeenCalledWith([
                { type: "area", key: "series 1"},
                { type: "line", key: "series 2", hidden: false }
            ]);
        });
    });
});
