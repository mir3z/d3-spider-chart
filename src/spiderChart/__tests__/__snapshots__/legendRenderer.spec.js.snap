// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`legend renderer renders correct legend cell for each hidden series 1`] = `
<g
  class="legend"
  transform="translate(50, 12)"
>
  <g
    class="legend-cell"
    transform="translate(0, 0)"
  >
    <text
      alignment-baseline="central"
      class="legend-cell__label"
      dominant-baseline="central"
      fill="#afafaf"
      x="2.5em"
    >
      series 1
    </text>
    <rect
      class="legend-cell__symbol"
      fill="#fff"
      height="1em"
      stroke="#afafaf"
      width="2em"
      y="-0.5em"
    />
  </g>
  <g
    class="legend-cell"
    transform="translate(0, 30)"
  >
    <text
      alignment-baseline="central"
      class="legend-cell__label"
      dominant-baseline="central"
      fill="#afafaf"
      x="2.5em"
    >
      series 2
    </text>
    <line
      class="legend-cell__symbol"
      stroke="#afafaf"
      stroke-width="4"
      x1="0"
      x2="2em"
    />
  </g>
  <g
    class="legend-cell"
    transform="translate(0, 60)"
  >
    <text
      alignment-baseline="central"
      class="legend-cell__label"
      dominant-baseline="central"
      fill="#afafaf"
      x="2.5em"
    >
      series 3
    </text>
    <circle
      class="legend-cell__symbol"
      cx="1em"
      fill="#fff"
      r="4px"
      stroke="#afafaf"
    />
  </g>
  <line
    class="line-decor"
    x1="0"
    x2="300"
    y1="-20"
    y2="-20"
  />
  <line
    class="line-decor"
    x1="0"
    x2="300"
    y1="80"
    y2="80"
  />
</g>
`;

exports[`legend renderer renders correct legend cell for each series 1`] = `
<g
  class="legend"
  transform="translate(50, 12)"
>
  <g
    class="legend-cell"
    transform="translate(0, 0)"
  >
    <text
      alignment-baseline="central"
      class="legend-cell__label"
      dominant-baseline="central"
      fill="#000"
      x="2.5em"
    >
      series 1
    </text>
    <rect
      class="legend-cell__symbol"
      fill="rgb(255, 0, 0)"
      height="1em"
      stroke="none"
      width="2em"
      y="-0.5em"
    />
  </g>
  <g
    class="legend-cell"
    transform="translate(0, 30)"
  >
    <text
      alignment-baseline="central"
      class="legend-cell__label"
      dominant-baseline="central"
      fill="#000"
      x="2.5em"
    >
      series 2
    </text>
    <line
      class="legend-cell__symbol"
      stroke="rgb(0, 255, 0)"
      stroke-width="4"
      x1="0"
      x2="2em"
    />
  </g>
  <g
    class="legend-cell"
    transform="translate(0, 60)"
  >
    <text
      alignment-baseline="central"
      class="legend-cell__label"
      dominant-baseline="central"
      fill="#000"
      x="2.5em"
    >
      series 3
    </text>
    <circle
      class="legend-cell__symbol"
      cx="1em"
      fill="rgb(0, 0, 255)"
      r="4px"
      stroke="none"
    />
  </g>
  <line
    class="line-decor"
    x1="0"
    x2="300"
    y1="-20"
    y2="-20"
  />
  <line
    class="line-decor"
    x1="0"
    x2="300"
    y1="80"
    y2="80"
  />
</g>
`;
